server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.routing.RecoverySource;
+import org.elasticsearch.cluster.routing.RecoverySource.SnapshotRecoverySource;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+
+/**
+ * An allocation decider that prevents relocation or allocation from nodes
+ * that might not be index compatible. If we relocate from a node that uses
+ * a newer index version than the node we relocate to this might cause {@link org.apache.lucene.index.IndexFormatTooNewException}
+ * on the lowest level since it might have already written segments that use a new postings format or codec that is not
+ * available on the target node.
+ */
+public class IndexVersionAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "index_version";
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (shardRouting.primary()) {
+            if (shardRouting.currentNodeId() == null) {
+                if (shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    // restoring from a snapshot - check that the node can handle the version
+                    return isSnapshotIndexVersionCompatible((SnapshotRecoverySource) shardRouting.recoverySource(), node, allocation);
+                } else {
+                    // existing or fresh primary on the node
+                    return allocation.decision(Decision.YES, NAME, "the primary shard is new or already existed on the node");
```

Comment:
This is not really true in stateless, where there is no "already existed on the node". I wonder if we need to record the use of an `IndexVersion` in metadata when allocating a shard, but suspect there could be other strategies too.

I think this is umodified by the PR and hence we could choose to ignore this for now, depending on the purpose of this PR.

I think we should at least check for the mapping updated version (which might be more of an edge case though).

Reply:
This PR is to replace `NodeVersionAllocationDecider` with equivalent functionality, but not referencing the node version. Extensions and enhancements for stateless would be separate work